Spanner/tests/Unit/DatabaseTest.php

```diff
@@ -1448,8 +1453,548 @@ public function testDBDatabaseRole()
+
+    public function testRunTransactionWithBlindCommit()
+    {
+        $this->stubCommit(false);
+        $this->refreshOperation($this->database, $this->connection->reveal());
+
+        $this->database->runTransaction(function ($t) {
+            $t->insert('Posts', [
+                'ID' => 10,
+                'title' => 'My New Post',
+                'content' => 'Hello World'
+            ]);
+            $t->commit();
+        }, ['tag' => self::TRANSACTION_TAG]);
+    }
+
+    public function testRunTransactionWithUnavailableErrorRetry()
+    {
+        $sql = $this->readArgs('sql');
+        $numOfRetries = 2;
+        $unavailable = new ServiceException('Unavailable', 14);
+        $result = $this->resultGenerator(true, self::TRANSACTION);
+
+        $it = 0;
+        // First call with ILB results in unavailable error.
+        // Second call also made with ILB, returns ResultSet.
+        $this->connection->executeStreamingSql(Argument::allOf(
+            Argument::withEntry('sql', $sql),
+            Argument::withEntry('transaction', self::BEGIN_RW_OPTIONS),
+            Argument::withEntry('requestOptions', [
+                'transactionTag' => self::TRANSACTION_TAG
+            ])
+        ))
+            ->shouldBeCalledTimes($numOfRetries)
+            ->will(function () use (&$it, $unavailable, $numOfRetries, $result) {
+                $it++;
+                if ($it < $numOfRetries) {
+                    throw $unavailable;
+                }
+                return $result;
+            });
+        $this->stubCommit();
+        $this->refreshOperation($this->database, $this->connection->reveal());
+
+        $this->database->runTransaction(function ($t) use ($sql) {
+            $t->execute($sql);
+            $t->commit();
+        }, ['tag' => self::TRANSACTION_TAG]);
+    }
+
+    public function testRunTransactionWithFirstUnavailableErrorRetry()
+    {
+        $sql = $this->readArgs('sql');
+        $unavailable = new ServiceException('Unavailable', 14);
+
+        // First call with ILB results in a transaction.
+        // Then the stream fails, Second call needs to use the
+        // transaction created by the first call.
+        $this->connection->executeStreamingSql(Argument::allOf(
+            Argument::withEntry('sql', $sql),
+            Argument::withEntry('transaction', self::BEGIN_RW_OPTIONS),
+            Argument::withEntry('requestOptions', [
+                'transactionTag' => self::TRANSACTION_TAG
+            ])
+        ))
+            ->shouldBeCalledTimes(1)
+            ->willreturn($this->resultGeneratorWithError());
+        $this->stubExecuteStreamingSql(['id' => self::TRANSACTION]);
+        $this->stubCommit();
+        $this->refreshOperation($this->database, $this->connection->reveal());
+
+        $this->database->runTransaction(function ($t) use ($sql) {
+            $result = $t->execute($sql);
+            foreach ($result as $res) {
+                continue;
+            }
```

Comment:
Do you wan to assert here in loop? Else remove?

Reply:
No need, removed.